### ISSUE TITLE
Current arRPC workaround not working

### ISSUE THREAD

**Issue by hideki2k02** [flathub/dev.vencord.Vesktop#11]
The current guide asks for the permission to access `xdg-run/discord-ipc-*` however that is currently not working and should be replaced by `xdg-run/discord-ipc-0`. Below are also a few of my findings:

Native Version:
- Native with Native apps works flawlessly (properly creates `discord-ipc-0` so no need to do it manually)
- Native with Flatpak apps requires extra perms `--filesystem=xdg-run/discord-ipc-0` (discord-ipc-* does not work/nobody runs two Discord clients)

Flatpak Version:
- Flatpak with Native apps requires symlink file `discord-ipc-0` to `/run/user/1000/.flatpak/dev.vencord.Vesktop/xdg-run/discord-ipc-0`
- Flatpak with Flatpak apps requires symlink file `discord-ipc-0` and `--filesystem=xdg-run/discord-ipc-0`

**Comment by lewisakura** [comment]
> nobody runs two Discord clients

this is a bad assumption to make as people can run multiple clients for many reasons, including trying different versions of the app, development, alt accounts, etc. and i personally do this for vencord development

this workaround is a variation on armcord's workaround, so it's interesting that it doesn't work when it seemingly works for them (or at the very least no one has flagged an issue with it).

feel free to open a PR to make the doc change, but keep the above two comments into account doing so

**Comment by hideki2k02** [comment]
> this is a bad assumption to make as people can run multiple clients for many reasons, including trying different versions of the app, development, alt accounts, etc. and i personally do this for vencord development

Yeah it was bad now that i think about it, but still you'll probably only be using one client for RPC. Otherwise you can use the wrapper mentioned [here](https://github.com/flathub/com.discordapp.Discord/wiki/Rich-Precense-(discord-rpc)) (Adapt it to your client/needs).

> no one has flagged an issue with it).
I've used ArmCord in the past and arRPC never worked, i've tried looking around but the answer is always "arRPC is weird and people dont know how it works"

I'll leave the Issue Open/wont make a PR for now so more people can test my solution and the "more experienced" can try to improve it.

**Comment by Wa1t5** [comment]
In my case I wrote this simple script and created a .desktop file to run it
for some reason if I symlink to xdg-run/discord-ipc-0 and give an app
access to it when I try to start the app for the second time it won't start
because bwrap will say the file is already symlinked (makes sense)
but for some weird reason when I hard-link the file bwrap just ignore it
and works normally, it would be fine if I just had to hard-link the file
but if I do that vesktop will say that the file is busy, so I have to delete
it and wait til vesktop start talking to the file

```sh
#!/usr/bin/bash

# Delete hard link
rm $XDG_RUNTIME_DIR/discord-ipc-0 &

# Open vesktop
flatpak run dev.vencord.Vesktop &

# Recreate hard link
sleep 10
ln -f $XDG_RUNTIME_DIR/.flatpak/dev.vencord.Vesktop/xdg-run/discord-ipc-0 $XDG_RUNTIME_DIR/discord-ipc-0
```

**Comment by FrankAlbella** [comment]
Can confirm I'm seeing the same issues in this thread using the Flatpak version. If I use `discord-ipc-*`, then the rich presence with another Flatpak app doesn't work. If I use `discord-ipc-0`, then it works the first time, but I cannot start that app again with the same bwrap error mentioned by Wa1t5. Maybe this is a bug in Flatpak?

To be clear my steps are:
1. symlink created through service at startup
2. Open Vesktop (Flatpak)
3. Open Ryujinx (Flatpak)
4. Everything works! (All apps open, RPC working)
5. Close Ryujinx
6. Attempt to reopen Ryujinx
7. Unable to open with the following error: `bwrap: Can't make symlink at /run/user/1000/discord-ipc-0: File exists`

**Comment by FrankAlbella** [comment]
Seems like Flatpak has a fix for this recently merged to main. flatpak/flatpak@a1bb185

**Comment by pparmy** [comment]
I can't seem to get RPC working despite the flatpak fixes; I'm on SteamOS
I used flatseal but couldn't get it functioning

**Comment by FrankAlbella** [comment]
> I can't seem to get RPC working despite the flatpak fixes; I'm on SteamOS I used flatseal but couldn't get it functioning

I don't think the fixes have made it to a stable release yet.

**Comment by devurandom** [comment]
The [bubblewrap changes](https://github.com/flatpak/flatpak/commit/a1bb18561fa3a220b1244ec871f85518a2eeb763) should be available in Flatpak 1.15.7 and newer, if your distro has that already.  (Update request for NixOS: https://github.com/NixOS/nixpkgs/issues/326906)

However, I could not get Rich Presence to work, not even on the first run. The [Steam Flatpak already seems to be set up](https://github.com/flathub/com.valvesoftware.Steam/blob/beta/com.valvesoftware.Steam.yml#L36) to [work with the Discord Flatpak](https://github.com/flathub/com.valvesoftware.Steam/blob/beta/steam_wrapper/steam_wrapper.py#L406-L415), but (a) the [Discord Flatpak sets up the sockets a little bit different from Vesktop](https://github.com/flathub/com.discordapp.Discord/blob/master/discord.sh#L3-L6), so the [Discord documentation](https://github.com/flathub/com.discordapp.Discord/wiki/Rich-Precense-(discord-rpc)#flatpak-applications) does not directly apply to Vesktop, and (b) the Steam Flatpak is not set up to access the Vesktop sockets.

**Comment by Wa1t5** [comment]
> The bubblewrap changes should be available in Flatpak 1.15.7 and newer, if your distro has that already.
> 
> However, I could not get Rich Presence to work, not even on the first run. The [Steam Flatpak already seems to be set up](https://github.com/flathub/com.valvesoftware.Steam/blob/beta/com.valvesoftware.Steam.yml#L36) to [work with the Discord Flatpak](https://github.com/flathub/com.valvesoftware.Steam/blob/beta/steam_wrapper/steam_wrapper.py#L406-L415), but (a) the [Discord Flatpak sets up the sockets a little bit different from Vesktop](https://github.com/flathub/com.discordapp.Discord/blob/master/discord.sh#L3-L6), so the [Discord documentation](https://github.com/flathub/com.discordapp.Discord/wiki/Rich-Precense-(discord-rpc)#flatpak-applications) does not directly apply to Vesktop, and (b) the Steam Flatpak is not set up to access the Vesktop sockets.

The last time I tested vesktop + steam my fix worked fine, I know it's a bit inconvenient to have to setup a .desktop file but it works most of the times and save the headache of having to configure it every time you launch vesktop

**Comment by Arcitec** [comment]
The current guide says:

> Add `--filesystem=xdg-run/discord-ipc-*` to the global Flatpak permissions

That is pure nonsense and does nothing at all.

`--filesystem=` can ONLY point at directories. It does NOTHING if you point it at files.

Furthermore, the directory to-be-mapped MUST exist when the Flatpak starts. If it doesn't exist, then it doesn't get mapped into the Flatpak (until the Flatpak is restarted) even if the host creates it later. There is a `:create` suffix if you want your Flatpak to auto create the host-directory if missing (which avoids that race condition).

You also cannot use wildcards in filesystem access, meaning that `discord-ipc-*` has zero effect. On top of the issue that you cannot map files, only directories.

arRPC attempts to create a file in `xdg-run`, by the way. But to create new files or directories, the Flatpak MUST have access to the PARENT directory of the new file/dir.

Anyway, this means that the only correct permission is:

`--filesystem=xdg-run`

This will then map the run-directory into the Flatpak, so that when arRPC creates the discord-ipc file, it will be written to the host... Where it can be seen and used by host apps, and even by other flatpaks (if those have access to the RPC path).

In case full access to the entirety of `xdg-run` is not permitted, you can do what the official Discord Flatpak does, and put the discord-ipc file in a subdirectory of that. They use the equivalent of `--filesystem=xdg-run/app/com.discordapp.Discord:create` and modified Discord to write its IPC socket there, and then they tell people to [symlink the contents of that directory to the real location on the host so that native apps can reach it](https://github.com/flathub/com.discordapp.Discord/wiki/Rich-Precense-(discord-rpc)#unsandboxed-applications), but they actually only bother with discord-ipc-0 which means their script breaks if Discord has to create another IPC socket. Then they also ask the maintainers of other Flatpaks to adapt themselves to that location if they want to connect to Rich Presence, by adding the Discord xdg-run subdirectory and then run a script inside their own Flatpaks which links the normal xdg-run/discord-ipc-0 (inside each 3rd party Flatpak) to the Discord Flatpak RPC location exported by the host. They provide a sample script for other Flatpak authors to run inside their own Flatpak sandboxes, but the code for it is very poorly written and breaks things (creates a symlink to nowhere if discord-ipc didn't already exist, and doesn't work at all if the Host doesn't have the Discord xdg-run subdirectory yet which in turn means that it doesn't get mapped into the other Flatpaks if you start discord AFTER the other Flatpak apps). Their suggested 3rd party scripts also have no integration with the native Discord's RPC, only the Flatpak Discord.

I am guessing the real reason they used a subdirectory is just for extra security and deeper sandboxing. But considering how poorly written their scripts are, they have not really thought about this properly so I wouldn't take anything they did as a "good reference".

If we are permitted to use `--filesystem=xdg-run` then we can avoid all of that junk and all of their poorly written scripts.

And other Flatpaks would only need access to xdg-run to see the socket. But that's assuming those Flatpaks haven't implemented the Discord Flatpak's terrible script which has the symlink and xdg-run subdirectory mapping issues I just mentioned. In that case, those Flatpaks expect the socket to be at `xdg-run/app/com.discordapp.Discord/discord-ipc-0`. So we would have to use that location!

Either way, all of this is super messy (much of it thanks to how badly the Discord flatpak did things). Which is why I use the native Vesktop. Besides, only the native one is able to scan for game processes.

**Comment by lewisakura** [comment]
You should file your complaints about the instructions over at the ArmCord Flatpak, this Flatpak is a derivative of theirs and their instructions: https://github.com/flathub/xyz.armcord.ArmCord. Like I said earlier in this thread it's weird that it seems to work for them and not us (most I can see in their repo that's obvious is a bug relating to not cleaning up the IPC socket?).

We cannot use `--filesystem=xdg-run`, it is too broad of a permission to include in a Flatpak and defeats a ton of the point of sandboxing in the first place, which is why it was meant to be a specific permission. We also are not going to adapt to a third-party Flatpak's methodology because that would be silly.

To be honest, RPC might be a fruitless endeavour at this point and we might just have to consider it an unsupported feature in the Flatpak. There are so many stupid hacks and modifications you have to do to our and other people's Flatpaks that lead to things just breaking or not working or, well, this.

**Comment by Arcitec** [comment]
Yeah, the way the Discord Flatpak did it has set the standard. I do not know if ANY other Flatpaks took their advice though (forcing themselves to look at `xdg-run/app/com.discordapp.Discord/discord-ipc-0`). (Edit: Yes, other Flatpaks use that standard already, I found references to it.)

But even with all its flaws, the most compatible way to do things would be exactly what the Discord Flatpak does: Moving the IPC to a `com.discordapp.Discord` xdg-run subdirectory:

https://github.com/flathub/com.discordapp.Discord

Because they have instructed the rest of the Flatpak world to integrate with that path. So if even one app does it, then everyone else would have to do it.

> To be honest, RPC might be a fruitless endeavour at this point and we might just have to consider it an unsupported feature in the Flatpak. There are so many stupid hacks and modifications you have to do to our and other people's Flatpaks that lead to things just breaking or not working or, well, this.

Exactly yeah... It's very telling that even the official Discord Flatpak hasn't figured out a bug-free solution after all this time and all of its users. I'd guess the main reason is that the people with a lot of expertise probably want privacy and don't care about "Rich Presence" integration, so the sloppy, broken scripts the Discord Flatpak suggested remained as the "standard".

As for ArmCord, I just had a look and they do not do things the way the Vesktop Flatpak does.

They instead do this:

- The ArmCord Flatpak writes to its own internal xdg-run location (inside the Flatpak filesystem).
- On the Host, they tell you to link `xdg-run/discord-ipc-0` to the internal Flatpak file, which indeed makes Native apps capable of talking to the Flatpak ArmCord's RPC server: https://github.com/flathub/xyz.armcord.ArmCord?tab=readme-ov-file#native-applications
- But in their instructions for other Flatpaks, they have COMPLETELY wrong instructions (the `discord-ipc-*` wildcard issue which does nothing, and the attempt to give access to a specific file rather than a directory (which is impossible in Flatpaks)), etc. Their Flatpak instructions do not work. On top of that, they tell people to create a symlink *inside* all *other* Flatpaks which "should point to the ArmCord Flatpak", which is not gonna happen because that's something the other Flatpak authors have to do in their own manifests and startup scripts, and is also something that should use the "standardized" path set by the original Discord Flatpak, not a per-Vencord/ArmCord path, because the chances of enough apps integrating with "The ArmCord path" is even lower than the chance of them integrating with "the official Discord Flatpak path".

If you want native app RPC integration, you can use a similar technique as what they recommended.

If you want other Flatpaks to talk to RPC, you need to do what the Discord Flatpak did, and use the same `xdg-run/app/com.discordapp.Discord/discord-ipc-0` paths. Preferably with `:create` permission suffix to ensure the path is created when Vesktop starts up, so that other Flatpaks can map it into their filesystems, to lower the risk of  app launch order race conditions with other Flatpaks (which is an issue even with the official Discord Flatpak).

I'll let ArmCord know about the issues. :)

Edit: I see that other people using ArmCord have already noticed that their instructions are incorrect: https://github.com/flathub/xyz.armcord.ArmCord/issues/54 and https://github.com/flathub/xyz.armcord.ArmCord/issues/53

**Comment by Vendicated** [comment]
I revised the RPC instructions a while ago, so I'll be closing this: https://github.com/flathub/dev.vencord.Vesktop#discord-rich-presence

If there's still anything wrong, just comment again or feel free to open a pull request
